ibis/pandas/core.py

```diff
@@ -554,3 +610,113 @@ def canonicalize_context(
         )
     return begin, end
+
+
+class TimeContextRelation(enum.Enum):
+    """ Enum to classify the relationship between two time contexts
+        Assume that we have two timecontext c1 (begin1, end1),
+        c2(begin2, end2):
+        SUBSET means c1 is a subset of c2, begin1 is greater than or equal to
+        begin2, and end1 is less than or equal to end2.
+        Likewise, SUPERSET means that begin1 is earlier than begin2, and end1
+        is later than end2.
+        If neither of the two contexts is a superset of each other, and they
+        share some time range in common, we called them OVERLAP.
+        And NONOVERLAP means the two contexts doesn't overlap at all, which
+        means end1 is earlier than begin2 or end2 is earlier than begin1
+    """
+
+    SUBSET = 0
+    SUPERSET = 1
+    OVERLAP = 2
+    NONOVERLAP = 3
+
+
+def compare_timecontext(cur_context: TimeContext, old_context: TimeContext):
+    """Compare two timecontext and return the relationship between two time
+    context (SUBSET, SUPERSET, OVERLAP, NONOVERLAP).
+
+    Parameters
+    ----------
+    cur_context: TimeContext
+    old_context: TimeContext
+
+    Returns
+    -------
+    result : Enum[TimeContextRelation]
+    """
+    begin, end = cur_context
+    old_begin, old_end = old_context
+    if old_begin <= begin and old_end >= end:
+        return TimeContextRelation.SUBSET
+    elif old_begin >= begin and old_end <= end:
+        return TimeContextRelation.SUPERSET
+    elif old_end < begin or end < old_begin:
+        return TimeContextRelation.NONOVERLAP
+    else:
+        return TimeContextRelation.OVERLAP
+
+
+def scope_item(op, result, timecontext):
+    """make a scope item to be set in scope
+
+    Scope is a dictionary mapping :class:`~ibis.expr.operations.Node`
+    subclass instances to concrete data, and the time context associate
+    with it(if any).
+
+    Parameters
+    ----------
+    op: ibis.expr.operations.Node, key in scope.
+    result : scalar, pd.Series, pd.DataFrame, concrete data.
+    timecontext: Optional[TimeContext], time context associate with the
+    result.
+
+    Returns
+    -------
+    scope_item : Dict, a key value pair that could merge into scope later
+    """
+    return {op: {'value': result, 'timecontext': timecontext}}
+
+
+def get_scope(scope, op, timecontext=None):
+    """ Given a op and timecontext, get result from scope
+    Parameters
+    ----------
+    scope: collections.Mapping
+    a dictionary mapping :class:`~ibis.expr.operations.Node`
+    subclass instances to concrete data, and the time context associate
+    with it(if any).
+    op: ibis.expr.operations.Node, key in scope.
+    timecontext: Optional[TimeContext]
+
+    Returns
+    -------
+    result : scalar, pd.Series, pd.DataFrame
+    """
+    if op not in scope:
+        return None
+    # for ops without timecontext
+    if timecontext is None:
```

Comment:
For consistency, we should perhaps use an timecontext place holder for the `timecontext is None` case

Reply:
Do you mean have an entry of {"value":None, "timecontext":None}?